From ffd9c24a8fe12fe8a74bc9446b7073a8e8eaef8c Mon Sep 17 00:00:00 2001
From: radomir-sebek <rasho.ieee@gmail.com>
Date: Wed, 2 May 2018 22:11:47 +0200
Subject: [PATCH 04/22] [Improvements] Check did npm install installed all
 required packages (#89)

* [Improvements] Check did npm install did installed all requered packages

* tiny fix
---
 modules/Utilities.js | 27 +++++++++++++++++++++++++++
 ot-node.js           |  5 ++++-
 package.json         |  1 +
 3 files changed, 32 insertions(+), 1 deletion(-)

diff --git a/modules/Utilities.js b/modules/Utilities.js
index ed20958..6cfeacd 100644
--- a/modules/Utilities.js
+++ b/modules/Utilities.js
@@ -73,6 +73,33 @@ class Utilities {
     }
 
     /**
+     * Check if all dependencies from package.json are installed
+     * @returns {Promise<any>} containing error array:
+     *   error: []            // when everything is OK, error array is empty
+     *   error: array,        // if not OK, array of logged errors is returned
+     */
+    static checkInstalledDependencies() {
+        return new Promise((resolve, reject) => {
+            // eslint-disable-next-line global-require
+            require('check-dependencies')({
+                packageManager: 'npm',
+                // eslint-disable-next-line no-template-curly-in-string
+                packageDir: '${__dirname}/../',
+                install: false,
+                scopeList: ['dependencies', 'devDependencies'],
+                verbose: false,
+            }).then((output) => {
+                if (!output.depsWereOk) {
+                    reject(output.error);
+                }
+                resolve(output.error);
+            }).catch((error) => {
+                reject(error);
+            });
+        });
+    }
+
+    /**
      * Returns winston logger
      * @returns {*} - log function
      */
diff --git a/ot-node.js b/ot-node.js
index d84e922..daf2ee5 100644
--- a/ot-node.js
+++ b/ot-node.js
@@ -40,12 +40,15 @@ class OTNode {
      */
     bootstrap() {
         try {
+            // check if all dependencies are installed
+            deasync(Utilities.checkInstalledDependencies());
+            log.info('npm modules dependences check done');
             // make sure arango database exists
             deasync(Utilities.checkDoesStorageDbExists());
             log.info('Storage database check done');
             // Checking root folder stucture
             Utilities.checkOtNodeDirStructure();
-            log.info('ot-node folder structure checked');
+            log.info('ot-node folder structure check done');
         } catch (err) {
             console.log(err);
         }
diff --git a/package.json b/package.json
index 3c101e0..d36c750 100644
--- a/package.json
+++ b/package.json
@@ -45,6 +45,7 @@
     "bn.js": "^4.11.8",
     "boscar": "^2.0.0",
     "bunyan": "^1.8.12",
+    "check-dependencies": "^1.1.0",
     "deasync-promise": "^1.0.1",
     "dotenv": "^5.0.1",
     "encoding-down": "^4.0.0",
-- 
2.13.5 (Apple Git-94)

